### PR TITLE
Open Tweet Assessment

### PR DESCRIPTION
**Pull request by r-b-ivaniv**
Screenshot 2024-01-29 at 4 22 36 PM
Screenshot 2024-01-27 at 2 39 46 PM

# Pros & Cons

## I'd like to share my thoughts on what I believe went well and areas I see for improvement. Spoiler alert: given more time, I could have enhanced certain aspects :)

# Pros:

1. **Multimodular Architecture:** Enhances scalability and improves parallel development. Explored the app with a perspective similar to Twitter.

2. **Efficient UI Layout:** Avoided using storyboards, XIB, autolayout, and UIStackViews (known to be slow). Instead, used the third-party library PinLayout (https://github.com/layoutBox/PinLayout, check the performance section).

3. **Efficient Cell Height Calculation:** Optimized cell height calculation using cell prototypes and manual calculations for better performance.

4. **Efficient Connection and Sorting of Tweets:** Utilized a constant dictionary (tweetMap) to efficiently connect and sort tweets on the fly.

5. **Efficient Image Loading + Cache:** Implemented efficient mechanisms for image loading and caching to enhance the overall performance of the app.

6. **UIViewController Screens with UICollectionView:** Provided more flexibility and customization opportunities. Implemented a custom complex animation, although I couldn't finish it due to time constraints.

7. **Handled Light/Dark Modes and Accessibility Edge Cases:** Ensured the app works well in different modes and considers accessibility scenarios.

8. **Added Unit Tests:** Created unit tests for core logic, specifically in TimelineViewModelTests.

9. **Error Handling:** Implemented error handling for edge cases to enhance the app's robustness.

10. **Maintained PRs for Each Task:** Maintained a clear and organized Git history with 7 pull requests, meeting both minimum and bonus requirements.

11. **Fixed Launch Image Performance Problem:** Resolved a significant performance issue related to the launch image by reducing its resolution.

# Cons:

1. **Super Simple UI:** Designed a minimalist UI without custom fonts, complex animations, etc.

2. **No Snapshot Testing:** Did not implement snapshot testing for screens.

3. **Simple Scale Animation for Cell Selection:** Implemented a basic scale animation for cell selection, without parallax or UIDynamics effects. Needed more time for additional enhancements.

4. **No Custom Transition Animation for Tweet Details:** Did not include a custom transition animation for tweet details. For instance, a custom transition layer to resize the selected tweet, slide it to the top, and show replies below. This feature required more time for implementation.

## Ideas:

- in real world for twitter like data structre core data/SwiftData could be a great fit (relationships, effiecient)
- Using swiftUI could give good performance boost for development

### CODE REVIEW

File: Documentation.docc/Pros & Cons.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+7. **Handled Light/Dark Modes and Accessibility Edge Cases:** Ensured the app works well in different modes and considers accessibility scenarios.
+
+8. **Added Unit Tests:** Created unit tests for core logic, specifically in TimelineViewModelTests.
+
+9. **Error Handling:** Implemented error handling for edge cases to enhance the app's robustness.
+
+10. **Maintained PRs for Each Task:** Maintained a clear and organized Git history with 7 pull requests, meeting both minimum and bonus requirements.
+
+11. **Fixed Launch Image Performance Problem:** Resolved a significant performance issue related to the launch image by reducing its resolution.
```
Comment (r-b-ivaniv) [inline Documentation.docc/Pros & Cons.md:29]: Screenshot 2024-01-27 at 8 16 14 PM

as you can see that launch icon have pretty large resolution and was no adopted for different screen size, but not worry, I fixed it :)

File: Features/Sources/Feed/Coordinator/FeedFlowCoordinator.swift [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+//  FeedFlowCoordinator.swift
+//
+//
+//  Created by Roman Ivaniv on 2024-01-26.
+//
+
+import UIKit
+import CoreModule
+
+public final class FeedFlowCoordinator: Coordinator {
```
Comment (r-b-ivaniv) [inline Features/Sources/Feed/Coordinator/FeedFlowCoordinator.swift:11]: Each feature should have it own FlowCoordinator

File: Features/Sources/Feed/Modules/Timeline/TimelineViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    }
+    
+    func connectReplies(in timeline: [TweetModel]) {
+        let tweetMap: [String: TweetModel] = Dictionary(uniqueKeysWithValues: timeline.map { ($0.id, $0) })
+
+        for tweet in timeline {
+            if let replyToId = tweet.inReplyTo, let replyTo = tweetMap[replyToId] {
+                // Connect the replies and sort them on the fly
+                replyTo.replies = (replyTo.replies ?? []) + [tweet]
+                replyTo.replies?.sort { $0.date < $1.date }
```
Comment (r-b-ivaniv) [inline Features/Sources/Feed/Modules/Timeline/TimelineViewModel.swift:67]: tried to achieve good performance for tweets connection logic. In `real world` we could do this in parallel while showing main screen, because we need replies only for tweet details screen, so we could show UI first and in parallel connect tweets

File: Features/Tests/FeedTests/TimelineViewModelTests.swift [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+final class TimelineViewModelTests: XCTestCase {
+    // TODO: more unit tests
+    func testConnectReplies() async throws {
+        let mockSevice = MockTimelineService()
+        let viewModel = TimelineViewModel(timelineService: mockSevice, onTweetAction: nil)
+        let timelineModel = try! await viewModel.timelineService.fetchTimeline()
+        let tweets = timelineModel.timeline
+        
+        viewModel.connectReplies(in: tweets)
```
Comment (r-b-ivaniv) [inline Features/Tests/FeedTests/TimelineViewModelTests.swift:12]: just some unit test to check connection logic for tweets

File: Features/Sources/Feed/Modules/Shared/Views/TweetCell.swift [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+            .right(Constants.inset)
+            .sizeToFit(.width)
+        
+        dateLabel.pin
+            .below(of: authorLabel, aligned: .left)
+            .right(Constants.inset)
+            .sizeToFit(.width)
+
+        contentTextView.pin
+            .below(of: [avatarImageView, dateLabel], aligned: .left)
```
Comment (r-b-ivaniv) [inline Features/Sources/Feed/Modules/Shared/Views/TweetCell.swift:129]: so as you can see content will be always below avatar and dateLabel (like lets image case when user had super long name that will take few lines)

File: Features/Sources/Feed/Modules/Thread/ThreadViewController.swift [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        
+        collectionView.register(TweetCell.self)
+        collectionView.registerSupplementaryView(SectionHeaderView.self, king: UICollectionView.elementKindSectionHeader)
+        
+        applySnapshot(for: viewModel.tweet)
+    }
+    
+    // MARK: - Private
+        
+    private func applySnapshot(for tweet: TweetModel) {
```
Comment (r-b-ivaniv) [inline Features/Sources/Feed/Modules/Thread/ThreadViewController.swift:57]: in real world I would move NSDiffableDataSourceSnapshot logic to viewModel later + added some tests, but to make it accurate and reasonable need to create some extra logic, unfortunately no time for this

File: InterfaceComponents/Sources/InterfaceComponents/Common/UIImageView+Load.swift [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    private static var imageCache = NSCache<NSURL, UIImage>()
+    private static var loadingResponses = [NSURL: [(UIImage?) -> Void]]()
+
+    func loadImage(from url: NSURL, placeholder: UIImage? = nil) {
+        // Set the placeholder image initially
+        self.image = placeholder
+
+        // Check for a cached image.
+        if let cachedImage = UIImageView.imageCache.object(forKey: url) {
+            self.image = cachedImage
```
Comment (r-b-ivaniv) [inline InterfaceComponents/Sources/InterfaceComponents/Common/UIImageView+Load.swift:20]: added cache image logic, just to not duplicate requests

File: Features/Sources/Feed/Modules/Shared/Views/TweetCell.swift [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+            // TODO: to get better performance we could sned request in willDisplayCell method
+            avatarImageView.loadImage(from: url as NSURL)
+        }
+        
+        setNeedsLayout()
+    }
+    
+    private func animateSelection(_ isSelected: Bool) {
+        // TODO: in real world custom animation transition to TweetViewController
+        UIView.animate(withDuration: Constants.animationDuration, animations: { [weak self] in
```
Comment (r-b-ivaniv) [inline Features/Sources/Feed/Modules/Shared/Views/TweetCell.swift:98]: as a quick solution just some simple scale animation for cell selection. In general i was focused more on core logic and project scalability and not on UI, which probably was a mistake 🥇

File: Features/Sources/Feed/Modules/Shared/Views/TweetCell.swift [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    func configure(tweet: TweetModel) {
+        authorLabel.text = tweet.author
+        dateLabel.text = tweet.date.iso8601FormattedDate()
+        contentTextView.text = tweet.content
+        
+        if let path = tweet.avatar, let url = URL(string: path) {
+            // TODO: to get better performance we could sned request in willDisplayCell method
+            avatarImageView.loadImage(from: url as NSURL)
```
Comment (r-b-ivaniv) [inline Features/Sources/Feed/Modules/Shared/Views/TweetCell.swift:90]: we still can achieve more efficient cell layout, by moving some logic around - like starting image download willDisplayCell

File: Features/Sources/Feed/Modules/Thread/ThreadViewController.swift [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+        case single
+        case replyTo
+        case replies
+        
+        var header: String? {
+            switch self {
+            case .replyTo:
+                return "In reply to:"
+            case .replies:
+                return "Replies:"
```
Comment (r-b-ivaniv) [inline Features/Sources/Feed/Modules/Thread/ThreadViewController.swift:109]: unfortunately didn't have time to add localization logic

File: Features/Sources/Feed/Modules/Shared/Views/TweetCell.swift [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+        
+        separatorView.pin
+            .below(of: contentTextView)
+            .marginTop(Constants.inset)
+            .height(Constants.separatorSize)
+            .horizontally()
+    }
+    
+    override func sizeThatFits(_ size: CGSize) -> CGSize {
+        return autoSizeThatFits(size, layoutClosure: layout)
```
Comment (r-b-ivaniv) [inline Features/Sources/Feed/Modules/Shared/Views/TweetCell.swift:142]: pretty effect logic to calculate cell height: I used cell prototype property (single instance) and use to calculate hight (autoSizeThatFits use manual calculation)
